[PyTorch] check isValidUnboxed() in the dispatcher

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51368 [PyTorch] Use real if constexpr behind macro in hot template
* #51367 [PyTorch] Debug-gate static_assert in KernelFunction::makeFromUnboxedFunctor
* #51319 [PyTorch] Remove reference_cast in make_boxed_from_unboxed_functor
* #51463 [PyTorch] Fix return value of IValue::to for Tensor/String
* #51315 [PyTorch][easy] Fix TODOs in CppFunction constructors
* #51313 [PyTorch] Reduce template expansion in call_functor_with_args_from_stack
* #51307 [PyTorch] Remove always-empty string args to inferFunctionSchemaFromFunctor
* #51293 [PyTorch] IWYU in torch/csrc/utils/future.h
* #51269 [PyTorch] Outline OperatorEntry::assertSignatureIsCorrect fail path
* **#51247 [PyTorch] check isValidUnboxed() in the dispatcher**
* #51165 [PyTorch] Pass TensorOptions by value
* #51245 [PyTorch] Add C10_ALWAYS_INLINE to critical dispatcher paths
* #51163 [PyTorch] Refactor Dispatcher to inline less code in fast path

See code comment for explanation.

This measures neutral compared to the previous diff with `perf stat` when running on a
benchmark that calls empty in a loop. I think that we should commit it
anyway because:
1) I have previously seen it make a difference when applied earlier in
the stack.
2) This makes sense both on principle and via inspecting output
assembly: we avoid having to touch the boxed kernel at all (usually)
and instead use the unboxed kernel for both the validity check in
`OperatorEntry::lookup` and the actual `KernelFunction::call`.

Differential Revision: [D26113650](https://our.internmc.facebook.com/intern/diff/D26113650/)

swolchok committed:
[PyTorch] check isValidUnboxed() in the dispatcher

See code comment for explanation.

This measures neutral compared to the previous diff with `perf stat` when running on a
benchmark that calls empty in a loop. I think that we should commit it
anyway because:
1) I have previously seen it make a difference when applied earlier in
the stack.
2) This makes sense both on principle and via inspecting output
assembly: we avoid having to touch the boxed kernel at all (usually)
and instead use the unboxed kernel for both the validity check in
`OperatorEntry::lookup` and the actual `KernelFunction::call`.

Differential Revision: [D26113650](https://our.internmc.facebook.com/intern/diff/D26113650/)

[ghstack-poisoned]

swolchok committed:
actually add the code comment on "[PyTorch] check isValidUnboxed() in the dispatcher"


See code comment for explanation.

This measures neutral compared to the previous diff with `perf stat` when running on a
benchmark that calls empty in a loop. I think that we should commit it
anyway because:
1) I have previously seen it make a difference when applied earlier in
the stack.
2) This makes sense both on principle and via inspecting output
assembly: we avoid having to touch the boxed kernel at all (usually)
and instead use the unboxed kernel for both the validity check in
`OperatorEntry::lookup` and the actual `KernelFunction::call`.

Differential Revision: [D26113650](https://our.internmc.facebook.com/intern/diff/D26113650/)

[ghstack-poisoned]

ezyang reviewed (approved):
This is fine. But @smessmer, now that everything is c10 full, don't we guarantee boxed is valid everywhere? So we ought to tighten the invariants here.

swolchok committed:
Update on "[PyTorch] check isValidUnboxed() in the dispatcher"


See code comment for explanation.

This measures neutral compared to the previous diff with `perf stat` when running on a
benchmark that calls empty in a loop. I think that we should commit it
anyway because:
1) I have previously seen it make a difference when applied earlier in
the stack.
2) This makes sense both on principle and via inspecting output
assembly: we avoid having to touch the boxed kernel at all (usually)
and instead use the unboxed kernel for both the validity check in
`OperatorEntry::lookup` and the actual `KernelFunction::call`.

Differential Revision: [D26113650](https://our.internmc.facebook.com/intern/diff/D26113650/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] check isValidUnboxed() in the dispatcher"


See code comment for explanation.

This measures neutral compared to the previous diff with `perf stat` when running on a
benchmark that calls empty in a loop. I think that we should commit it
anyway because:
1) I have previously seen it make a difference when applied earlier in
the stack.
2) This makes sense both on principle and via inspecting output
assembly: we avoid having to touch the boxed kernel at all (usually)
and instead use the unboxed kernel for both the validity check in
`OperatorEntry::lookup` and the actual `KernelFunction::call`.

Differential Revision: [D26113650](https://our.internmc.facebook.com/intern/diff/D26113650/)

[ghstack-poisoned]
